File: src/types/requestWorkflow.ts
import type { AuthorPreview } from './author';

export type WorkflowStatus =
  | 'draft'
  | 'new'
  | 'triage'
  | 'waiting_info'
  | 'in_progress'
  | 'review'
  | 'completed'
  | 'cancelled'
  | 'archived';

export type TaskStatus =
  | 'todo'
  | 'in_progress'
  | 'blocked'
  | 'review'
  | 'done'
  | 'cancelled';

export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent';

export type HistoryEventType =
  | 'request_created'
  | 'request_updated'
  | 'request_status_changed'
  | 'task_created'
  | 'task_updated'
  | 'task_status_changed'
  | 'task_assignee_changed'
  | 'task_completed'
  | 'request_completed'
  | 'request_cancelled';

export interface RequestTaskAssigneePreview {
  id: string;
  first_name?: string | null;
  last_name?: string | null;
  middle_name?: string | null;
  email?: string | null;
}

export interface RequestTaskDTO {
  id: string;
  request_id: string;
  title: string;
  description?: string | null;
  status: TaskStatus;
  assignee_id?: string | null;
  assignee?: RequestTaskAssigneePreview | null;
  created_by_id?: string | null;
  created_by?: AuthorPreview | null;
  priority: TaskPriority;
  due_date?: string | null;
  is_required: boolean;
  order_index: number;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
}

export interface RequestHistoryActor {
  id: string;
  first_name?: string | null;
  last_name?: string | null;
  middle_name?: string | null;
  email?: string | null;
}

export interface RequestHistoryEventDTO {
  id: string;
  request_id: string;
  actor_id?: string | null;
  actor?: RequestHistoryActor | null;
  type: HistoryEventType | string;
  payload?: Record<string, unknown> | null;
  created_at: string;
}

export interface WorkflowStatusSuggestion {
  suggested_status: WorkflowStatus;
  reason: string;
}

export const WORKFLOW_STATUS_LABELS: Record<WorkflowStatus, string> = {
  draft: 'Черновик',
  new: 'Новая',
  triage: 'На разборе',
  waiting_info: 'Ожидает информации',
  in_progress: 'В работе',
  review: 'На проверке',
  completed: 'Завершена',
  cancelled: 'Отменена',
  archived: 'В архиве',
};

export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  todo: 'К выполнению',
  in_progress: 'В работе',
  blocked: 'Заблокирована',
  review: 'На проверке',
  done: 'Выполнена',
  cancelled: 'Отменена',
};

export const TASK_PRIORITY_LABELS: Record<TaskPriority, string> = {
  low: 'Низкий',
  medium: 'Средний',
  high: 'Высокий',
  urgent: 'Срочный',
};

export const WORKFLOW_STATUS_COLORS: Record<WorkflowStatus, string> = {
  draft: 'default',
  new: 'blue',
  triage: 'cyan',
  waiting_info: 'orange',
  in_progress: 'processing',
  review: 'purple',
  completed: 'success',
  cancelled: 'error',
  archived: 'default',
};

export const TASK_STATUS_COLORS: Record<TaskStatus, string> = {
  todo: 'default',
  in_progress: 'processing',
  blocked: 'error',
  review: 'purple',
  done: 'success',
  cancelled: 'default',
};

export const TASK_PRIORITY_COLORS: Record<TaskPriority, string> = {
  low: 'default',
  medium: 'blue',
  high: 'orange',
  urgent: 'red',
};

File: backend/app/schemas/request_workflow.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.constants import request_workflow as W
from app.schemas.user import PublicAuthorResponse


class PatchWorkflowStatusPayload(BaseModel):
    status: str = Field(..., min_length=1)
    comment: str | None = Field(default=None, max_length=2000)
    force_complete: bool = False

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        if value not in W.WORKFLOW_STATUSES:
            raise ValueError(f"Invalid workflow status: {value}")
        return value


class RequestTaskAssigneePreview(BaseModel):
    id: str
    first_name: str | None = None
    last_name: str | None = None
    middle_name: str | None = None
    email: str | None = None


class RequestTaskResponse(BaseModel):
    id: str
    request_id: str
    title: str
    description: str | None = None
    status: str
    assignee_id: str | None = None
    assignee: RequestTaskAssigneePreview | None = None
    created_by_id: str | None = None
    created_by: PublicAuthorResponse | None = None
    priority: str
    due_date: str | None = None
    is_required: bool
    order_index: int
    created_at: str
    updated_at: str
    completed_at: str | None = None


class CreateRequestTaskPayload(BaseModel):
    title: str = Field(..., min_length=3, max_length=500)
    description: str | None = Field(default=None, max_length=5000)
    status: str = W.TASK_TODO
    assignee_id: str | None = None
    priority: str = W.PRIORITY_MEDIUM
    due_date: str | None = None
    is_required: bool = False
    order_index: int | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        if value not in W.TASK_STATUSES:
            raise ValueError(f"Invalid task status: {value}")
        return value

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, value: str) -> str:
        if value not in W.PRIORITIES:
            raise ValueError(f"Invalid priority: {value}")
        return value


class UpdateRequestTaskPayload(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=500)
    description: str | None = Field(default=None, max_length=5000)
    status: str | None = None
    assignee_id: str | None = None
    priority: str | None = None
    due_date: str | None = None
    is_required: bool | None = None
    order_index: int | None = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str | None) -> str | None:
        if value is not None and value not in W.TASK_STATUSES:
            raise ValueError(f"Invalid task status: {value}")
        return value

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, value: str | None) -> str | None:
        if value is not None and value not in W.PRIORITIES:
            raise ValueError(f"Invalid priority: {value}")
        return value


class PatchRequestTaskStatusPayload(BaseModel):
    status: str = Field(..., min_length=1)

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        if value not in W.TASK_STATUSES:
            raise ValueError(f"Invalid task status: {value}")
        return value


class PatchRequestTaskAssigneePayload(BaseModel):
    assignee_id: str | None = None


class RequestHistoryActorResponse(BaseModel):
    id: str
    first_name: str | None = None
    last_name: str | None = None
    middle_name: str | None = None
    email: str | None = None


class RequestHistoryEventResponse(BaseModel):
    id: str
    request_id: str
    actor_id: str | None = None
    actor: RequestHistoryActorResponse | None = None
    type: str
    payload: dict[str, Any] | None = None
    created_at: str


class WorkflowStatusSuggestion(BaseModel):
    suggested_status: str
    reason: str
